src/devices/usbCamera/unittest/testUltraPytonRunning.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2006-2021 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+//# @author Luca Tricerri <luca.tricerri@iit.it>
+
+#include "../linux/UltraPythonCameraHelper.h"
+#include "CApiMock.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include <chrono>
+#include <linux/v4l2-controls.h>
+#include <thread>
+
+using namespace std::chrono_literals;
+using namespace testing;
+
+TEST(UltraPython, step_base_ok) {
+  // given
+  InterfaceFoCApiMock *interface = new InterfaceFoCApiMock();
+  UltraPythonCameraHelper helper(interface);
+
+  struct v4l2_buffer buf;
+  buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  buf.memory = V4L2_MEMORY_MMAP;
+  buf.index = 1;
+  buf.flags = 0;
+  buf.sequence = 0;
+  buf.bytesused=10000;
+  EXPECT_CALL(*interface, select_c(_, _, nullptr, nullptr, _))
+      .WillOnce(Return(1));
+  EXPECT_CALL(*interface, xioctl_v4l2(_, VIDIOC_DQBUF, _)).WillOnce(DoAll(SetArgPointee<2>(buf), Return(1)));
+  EXPECT_CALL(*interface, xioctl(_, VIDIOC_QBUF, _)).WillOnce(Return(1));
+  
+  // when
+  bool out = helper.step();
+
+  // then
+  EXPECT_TRUE(out);
+
+  delete interface;
+}
+
+TEST(UltraPython, step_base_error1) {
+  // given
+  InterfaceFoCApiMock *interface = new InterfaceFoCApiMock();
+  UltraPythonCameraHelper helper(interface);
+
+  struct v4l2_buffer buf;
+  buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  buf.memory = V4L2_MEMORY_MMAP;
+  buf.index = 1;
+  buf.flags = V4L2_BUF_FLAG_ERROR;
+  buf.sequence = 0;
+  buf.bytesused=10000;
+  EXPECT_CALL(*interface, select_c(_, _, nullptr, nullptr, _))
+      .WillOnce(Return(1));
+  EXPECT_CALL(*interface, xioctl_v4l2(_, VIDIOC_DQBUF, _)).WillOnce(DoAll(SetArgPointee<2>(buf), Return(1)));
+  EXPECT_CALL(*interface, xioctl(_, VIDIOC_QBUF, _)).Times(0);
+  
+  // when
+  bool out = helper.step();
+
+  // then
+  EXPECT_FALSE(out);
+
+  delete interface;
+}
+
+TEST(UltraPython, step_base_timeout_ko) {
+  // given
+  InterfaceFoCApiMock *interface = new InterfaceFoCApiMock();
+  UltraPythonCameraHelper helper(interface);
+
+  EXPECT_CALL(*interface, select_c(_, _, nullptr, nullptr, _))
+      .WillOnce(Return(0));
+
+  // when
+  bool out = helper.step();
+
+  // then
+  EXPECT_FALSE(out);
+
+  delete interface;
+}
+
+TEST(UltraPython, step_base_error_ko) {
+  // given
+  InterfaceFoCApiMock *interface = new InterfaceFoCApiMock();
+  UltraPythonCameraHelper helper(interface);
+
+  EXPECT_CALL(*interface, select_c(_, _, nullptr, nullptr, _))
+      .WillOnce(Return(-1));
+
+  // when
+  bool out = helper.step();
+
+  // then
+  EXPECT_FALSE(out);
+
+  delete interface;
+}
```

Comment:
This comment is for all the unit test you added.

The unit test in yarp are in the test folder(maybe you can put it there https://github.com/robotology/yarp/tree/master/tests/libYARP_dev, we have to ask to @drdanz)
3. In YARP we use CATCH2 for unit test(see [ref](https://github.com/catchorg/Catch2)) I would stick to it and not introducing a different test suite(gtest).

Reply:
Relating to UnitTest they can not be written with other than gtest/gmock. This because the code deeply use Kernel API that should be moked (with gmock so gtest is mandatory). I suggest keeping these irregular UnitTests in this folder to avoid problems with the official tests.

Reply:
If this is viable/agreed, then we need to provide a CMake infrastructure that can detect/enable the GTest part and the corresponding unit tests smoothly and transparently w/o conflicting w/ the remaining test suite.

Reply:
However, there is a flag in the CMake. If activated the CMake will download the gtest and gmock library without problems. The alternative is to remove the tests but in my opinion is a pity.

Reply:
I see, there is the possibility to use catch2 with an external mock library(here [the example](https://github.com/sinojelly/catch2-with-mockcpp) and here is the [mockcpp library](https://github.com/sinojelly/mockcpp)) or catch with gmock(see https://github.com/matepek/catch2-with-gmock), here is the related issue https://github.com/catchorg/Catch2/issues/1659. I think that is a thing to be discussed with @drdanz

Reply:
Yes... but do not forget we will need gmock. So I do not see any advantage to use a mixed and potentially problematic test architecture. However I will follow @drdanz @Nicogene indications.

Reply:
I would like to discuss in a 30min meeting what's the most convenient way to proceed, I don't like either to include these handcrafted solutions, but I don't know if we can integrate the gtest in the current test suite that uses catch. If the test cannot run in the CI I think it is pointless to have this test, because we risk forgetting its existence.

Reply:
Will you organize the meeting? @Nicogene